working/0723-static-extensions/feature-specification-variant1.md

```diff
@@ -0,0 +1,467 @@
+type parameter scope, if any, and otherwise the library scope.
+
+Static members in a static extension are subject to the same static
+analysis as static members in other declarations.
+
+A factory constructor in a static extension introduces scopes in the same
+way as other factory constructor declarations. The return type of the
+factory constructor is the constructor return type of the static
+extension *(that is, the type in the `on` clause)*.
+
+Type variables of a static extension `E` are in scope in static member
+declarations in `E`, but any reference to such type variables in a static
+member declaration is a compile-time error. *The same rule applies for
+static members of classes, mixins, etc.*
+
+#### Static extension accessibility
+
+A static extension declaration _D_ is _accessible_ if _D_ is declared in
+the current library, or if _D_ is imported and not hidden.
+
+*In particular, it is accessible even in the case where there is a name
+clash with another locally declared or imported declaration with the same
+name.*
+
+#### Invocation of a static member of a static extension
+
+An _explicitly resolved invocation_ of a static member of a static
+extension named `E` is an expression of the form `E.m()` (or any other
+member access, *e.g., `E.m`, `E.m = e`, etc*), where `m` is a static member
+declared by `E`.
+
+*This can be used to invoke a static member of a specific static extension
+in order to manually resolve a name clash.*
+
+A static member invocation on a class `C`, of the form `C.m()` (or any
+other member access), is resolved by looking up static members in `C` named
+`m` and looking up static members of every accessible static extension with
+on-class `C` and a member named `m`.
+
+If `C` contains such a declaration then the expression is an invocation of
+that static member of `C`, with the same static analysis and dynamic
+semantics as before the introduction of this feature.
+
+Otherwise, an error occurs if no declarations named `m` or more than one
+declaration named `m` were found. *They would necessarily be declared in
+static extensions.*
+
+Otherwise, the invocation is resolved to the given static member
+declaration in a static extension named `Ej`, and the invocation is treated
+as `Ej.m()` *(this is an explicitly resolved invocation, which is specified
+above)*.
+
+#### The instantiated constructor return type of a static extension
+
+We associate a static extension declaration _D_ named `E` with formal type
+parameters `X1 extends B1 .. Xs extends Bs` and an actual type argument
+list `T1 .. Ts` with a type known as the _instantiated constructor return
+type of_ _D_ _with type arguments_ `T1 .. Ts`.
+
+When a static extension declaration _D_ named `E` has an on-clause which
+denotes a non-generic class `C`, the instantiated constructor return type
+is `C`, for any list of actual type arguments.
+
+*It is not very useful to declare a type parameter of a static extension
```

Comment:
I don't think this is true.  The parameters to the constructor (and the body) can use the type parameter.  Something like:
```dart
extension TypeKeyedMap<T> on Map<Type, Object> {
  Map.singleton(T x) => {T : x};
  void addByType<S>(S x) => this[S] = x;
}
```

Reply:
That's a bad use of the extension's type argument.
If we had generic constructors, it should be `Map.singleton<T>(T x) => {T: x}`.
Using an *extension* type parameter for that is a hack.

Not saying it won't work, or can't be useful, but it's also quite ugly.
Just use a static function. (Also, constructor should be `factory`.)

Reply:
> That's a (IMO) bad use of the extension's type argument.

Not sure what you're saying here.  Are you saying this should be an error?  I'm just trying to flesh out what the feature is.